File: utils/pinyin.py
from __future__ import annotations

import re


_TONE_MARKS = {
    "a": ["a", "ā", "á", "ǎ", "à"],
    "e": ["e", "ē", "é", "ě", "è"],
    "i": ["i", "ī", "í", "ǐ", "ì"],
    "o": ["o", "ō", "ó", "ǒ", "ò"],
    "u": ["u", "ū", "ú", "ǔ", "ù"],
    "ü": ["ü", "ǖ", "ǘ", "ǚ", "ǜ"],
}


_VOWELS = set("aeiouü")

_BREVE_TO_CARON = str.maketrans(
    {
        "ă": "ǎ",
        "Ă": "Ǎ",
        "ĕ": "ě",
        "Ĕ": "Ě",
        "ĭ": "ǐ",
        "Ĭ": "Ǐ",
        "ŏ": "ǒ",
        "Ŏ": "Ǒ",
        "ŭ": "ǔ",
        "Ŭ": "Ǔ",
    }
)


def numbers_to_tone_marks(pinyin: str) -> str:
    """
    Convert numbered pinyin (ni3, wo3) to tone marks (nǐ, wǒ).

    Handles:
    - whitespace separated syllables
    - v / u: for ü
    - neutral tone (5 or 0) -> no diacritic
    """
    s = (pinyin or "").strip()
    if not s:
        return s

    # Normalize ü inputs.
    s = s.replace("u:", "ü").replace("U:", "Ü").replace("v", "ü").replace("V", "Ü")

    parts = re.split(r"(\s+)", s)
    out: list[str] = []
    for part in parts:
        if part.isspace():
            out.append(part)
            continue
        out.append(_convert_syllable_token(part))
    return "".join(out).translate(_BREVE_TO_CARON)


def _convert_syllable_token(token: str) -> str:
    # Keep punctuation around syllables.
    m = re.match(r"^([^A-Za-züÜ]*)([A-Za-züÜ]+)([0-5]?)([^A-Za-züÜ]*)$", token)
    if not m:
        return token

    pre, core, tone_s, post = m.groups()
    if not tone_s:
        return token

    tone = int(tone_s)
    if tone in (0, 5):
        return f"{pre}{core}{post}"

    core_l = core.lower()
    core_l = core_l.replace("u:", "ü").replace("v", "ü")

    # Choose which vowel to mark (pinyin rules):
    # - if a or e exists, mark the first of those
    # - else if "ou" exists, mark o
    # - else mark the last vowel
    idx = _vowel_index_to_mark(core_l)
    if idx is None:
        return f"{pre}{core}{post}"

    marked = _apply_tone(core_l, idx, tone)
    # Restore capitalization heuristically (rare in our app).
    if core and core[0].isupper():
        marked = marked[0].upper() + marked[1:]
    return f"{pre}{marked}{post}"


def _vowel_index_to_mark(s: str) -> int | None:
    for v in ("a", "e"):
        i = s.find(v)
        if i != -1:
            return i
    i = s.find("ou")
    if i != -1:
        return i  # mark 'o'
    vowel_positions = [i for i, ch in enumerate(s) if ch in _VOWELS]
    return vowel_positions[-1] if vowel_positions else None


def _apply_tone(s: str, idx: int, tone: int) -> str:
    ch = s[idx]
    if ch not in _TONE_MARKS:
        return s
    rep = _TONE_MARKS[ch][tone]
    return s[:idx] + rep + s[idx + 1 :]

File: frontend/src/lib/pinyin.ts
/**
 * Pinyin normalization for writing test validation.
 * Converts tone numbers (lao3shi1) to tone marks (lǎoshī) so we can compare
 * user input (either format) with the correct answer (stored as numbers).
 */

const TONE_MARKS: Record<string, string[]> = {
  a: ["a", "ā", "á", "ǎ", "à"],
  e: ["e", "ē", "é", "ě", "è"],
  i: ["i", "ī", "í", "ǐ", "ì"],
  o: ["o", "ō", "ó", "ǒ", "ò"],
  u: ["u", "ū", "ú", "ǔ", "ù"],
  ü: ["ü", "ǖ", "ǘ", "ǚ", "ǜ"],
};

const VOWELS = new Set("aeiouü");

/** Map tone-mark vowels to base vowels for comparison (strip tones). */
const TONE_MARK_TO_BASE: Record<string, string> = {
  ā: "a", á: "a", ǎ: "a", à: "a",
  ē: "e", é: "e", ě: "e", è: "e",
  ī: "i", í: "i", ǐ: "i", ì: "i",
  ō: "o", ó: "o", ǒ: "o", ò: "o",
  ū: "u", ú: "u", ǔ: "u", ù: "u",
  ǖ: "ü", ǘ: "ü", ǚ: "ü", ǜ: "ü",
};

/**
 * Find which vowel index in the syllable gets the tone mark (pinyin rules).
 * - If 'a' or 'e' exists, mark the first of those.
 * - Else if "ou" exists, mark 'o'.
 * - Else mark the last vowel.
 */
function vowelIndexToMark(s: string): number | null {
  const lower = s.toLowerCase();
  for (const v of ["a", "e"]) {
    const i = lower.indexOf(v);
    if (i !== -1) return i;
  }
  const ou = lower.indexOf("ou");
  if (ou !== -1) return ou;
  let last = -1;
  for (let i = 0; i < lower.length; i++) {
    if (VOWELS.has(lower[i])) last = i;
  }
  return last >= 0 ? last : null;
}

function applyTone(s: string, idx: number, tone: number): string {
  const ch = s[idx].toLowerCase();
  const replacement = TONE_MARKS[ch]?.[tone];
  if (!replacement) return s;
  return s.slice(0, idx) + replacement + s.slice(idx + 1);
}

/**
 * Convert one syllable token (e.g. "lao3", "shi1") to tone marks.
 * Token = letters + optional tone digit 0-5.
 */
function convertSyllableToken(token: string): string {
  const trimmed = token.trim();
  const match = trimmed.match(/^([a-zA-ZüÜ]*)([0-5])?$/);
  if (!match) return trimmed;
  const [, core, toneStr] = match;
  if (!core || !toneStr) return trimmed;
  const tone = parseInt(toneStr, 10);
  if (tone === 0 || tone === 5) return core; // neutral

  let coreNorm = core.toLowerCase().replace(/v/g, "ü");
  const idx = vowelIndexToMark(coreNorm);
  if (idx === null) return core;
  return applyTone(coreNorm, idx, tone);
}

/**
 * Convert numbered pinyin to tone marks (e.g. lao3shi1 → lǎoshī).
 * Splits by syllable boundaries: letter run + digit 1-4 (or 0/5 for neutral).
 * If the string contains "(", only the part before "(" is converted; the rest is reattached as-is (e.g. wa2(wa) → wá(wa)).
 */
function numbersToToneMarksCore(s: string): string {
  if (!s) return "";
  const syllableRegex = /[a-zü]+[1-4]|[a-zü]+[05]?/gi;
  const tokens = s.match(syllableRegex) || [];
  return tokens.map(convertSyllableToken).join("");
}

export function numbersToToneMarks(pinyin: string): string {
  if (!pinyin || typeof pinyin !== "string") return "";
  const s = pinyin.trim().replace(/u:/gi, "ü").replace(/v/gi, "ü");
  const parenIdx = s.indexOf("(");
  if (parenIdx !== -1) {
    const pinyinPart = s.slice(0, parenIdx);
    const parenPart = s.slice(parenIdx);
    return numbersToToneMarksCore(pinyinPart) + parenPart;
  }
  return numbersToToneMarksCore(s);
}

/**
 * Convert tone-mark vowels to base vowels (ā→a, ē→e, etc.) so comparison is tone-agnostic.
 */
export function stripToneMarks(pinyin: string): string {
  if (!pinyin || typeof pinyin !== "string") return "";
  let s = pinyin;
  for (const [accented, base] of Object.entries(TONE_MARK_TO_BASE)) {
    s = s.split(accented).join(base);
  }
  return s;
}

/**
 * Normalize pinyin for comparison:
 * 1. Convert tone numbers in input to tone marks.
 * 2. Strip tone marks (convert to base vowels) on both user and dataset side.
 * 3. Lower case, no spaces, NFC.
 * Correct answer in the report is shown as original dataset pinyin (with tone marks), not this normalized form.
 */
export function normalizePinyinForComparison(input: string): string {
  if (!input || typeof input !== "string") return "";
  const trimmed = input.trim().toLowerCase().replace(/\s/g, "");
  if (!trimmed) return "";
  const withToneMarks = /[1-4]/.test(trimmed)
    ? numbersToToneMarks(trimmed).toLowerCase()
    : trimmed;
  return stripToneMarks(withToneMarks).normalize("NFC");
}
